### DISCUSSION TITLE
Possiblity of using an "unsafe" connection pool to speed up Lettuce performance under high concurrency

### DISCUSSION THREAD

**Discussion by yangty89** [redis/lettuce#2159] [Ideas]
Hi, I'am using JMH to test the performance of Lettuce under high concurrency (200 concurrent threads). The test is performed on a localhost Redis with a 8-processor CPU. The version of Lettuce is 6.1.6 and The Version of Redis is 6.2.6. 

I found that Lettuce performs well within a single shared connection - I got a QPS of about **60k**. But the performance drops slightly when I use a connection pool of 100 connection - I got a QPS of about **40k**. I think it's reasonable because when it comes to connection pooling, although we may take advantage of a multi-processor CPU to send requests on parallel, howerver, threads should obtain the connection first before sending a request, and the connection occupied by a thread could be released only when the thread has received the response from Redis. This differs a lot from single shared connection mode, in which all the threads could send requests to the single connection at same time. And from the view of Redis, it's some kind of pipelining, which has a positive impact on performance.

However, given that a Lettuce connection is thread safe for most Redis commands (blocking commands should be processed in dedicated connections), and Lettuce will dispose netty EventLoops according to CPU processor number, I think we may have a kind of "unsafe" connection pool - for example perhaps just a List of Lettuce connections, and let the thread choose randomly a connection for sending normal Redis requests. Thus we may take advantage of both multi-processor CPU and redis pipelining. I have made a test for this idea using 8 connections, and I got a QPS of about **100k**, which is nearly double the score I got for single thread mode and connection pooling mode :)

**Comment by mp911de** [comment]
Generally speaking, a connection count above the number of CPU cores is going to impose overhead. You already experienced an increase in throughput when reducing the connection count to 8. 

Lettuce connections are threadsafe and you do not need pooling unless you're using transactions or blocking commands.

**Reply by yangty89** [reply]
Hi mp911de, thanks for your reply :) Maybe I haven't described my idea clearly. In fact I'm thinking about maintaining just **a list of connections**, and every connection in this list could be shared by different threads at same time (given that Lettuce connection is thread safe for most Redis commands, and blocking commands should be processed in dedicated connection). It differs from traditional connection pool that a connection in the pool could only be owned by a single thread at one time. I have done some benchmarking for the connection list and the connection pool of both size 8 (number of my CPU processors), and I got a QPS of about **100k** for the connection list, and a QPS of about **40k** for the connection pool. So I think that maybe we could have something like this to improve the performance under high concurrency :) The test code and result is shown below:

```
@State(Scope.Benchmark)
public class CommonClientBenchmark {

    private RedisClient redisClient2;
    private RedisClient redisClient3;
    private List<StatefulRedisConnection<byte[], byte[]>> lettuceConnectionList;
    private GenericObjectPool<StatefulRedisConnection<byte[], byte[]>> lettucePool;
    private int processors;

    private String host = "127.0.0.1";
    private int port = 6379;

    @Setup
    public void setup() {
        System.out.println("*JMH SETUP CALLED*");
        
        // ** lettuce connection list **
        redisClient2 = RedisClient.create(redisURI);
        redisClient2.setOptions(ClientOptions.builder()
                .timeoutOptions(TimeoutOptions.builder().fixedTimeout(Duration.ofSeconds(10)).build()).build());
        processors = Runtime.getRuntime().availableProcessors(); // 8 on my laptop
    
        lettuceConnectionList = new ArrayList<>(processors);
        for (int i = 0; i < processors; ++i) {
            lettuceConnectionList.add(redisClient2.connect(ByteArrayCodec.INSTANCE));
        }

        // ** lettuce connection pool **
        redisClient3 = RedisClient.create(redisURI);
        redisClient3.setOptions(ClientOptions.builder()
                .timeoutOptions(TimeoutOptions.builder().fixedTimeout(Duration.ofSeconds(10)).build()).build());
        GenericObjectPoolConfig lettucePoolConfig = new GenericObjectPoolConfig();
        lettucePoolConfig.setMaxTotal(processors);
        lettucePoolConfig.setMaxIdle(processors);
        lettucePool = ConnectionPoolSupport
                .createGenericObjectPool(() -> redisClient3.connect(ByteArrayCodec.INSTANCE), lettucePoolConfig);
    }

    @Threads(200)
    @Fork(1)
    @Benchmark
    @BenchmarkMode(Mode.Throughput)
    @Warmup(iterations = 1, time = 30, timeUnit = TimeUnit.SECONDS)
    @Measurement(iterations = 2, time = 30, timeUnit = TimeUnit.SECONDS)
    public void lettucePoolSet() throws Exception {
        try (StatefulRedisConnection<byte[], byte[]> conn = lettucePool.borrowObject()) {
            conn.sync().set("lettu".getBytes(), "lettu".getBytes());
        }
    }

    @Threads(200)
    @Fork(1)
    @Benchmark
    @BenchmarkMode(Mode.Throughput)
    @Warmup(iterations = 1, time = 30, timeUnit = TimeUnit.SECONDS)
    @Measurement(iterations = 2, time = 30, timeUnit = TimeUnit.SECONDS)
    public void lettuceSetMultiConnection() {
        ThreadLocalRandom random = ThreadLocalRandom.current();
        lettuceConnectionList.get(random.nextInt(0, processors))
                .sync().set("lettu".getBytes(), "lettu".getBytes());
    }

}
```
```
Benchmark                                         Mode  Cnt       Score   Error  Units
CommonClientBenchmark.lettucePoolSet             thrpt    2   43752.025          ops/s
CommonClientBenchmark.lettuceSetMultiConnection  thrpt    2  102601.268          ops/s
```

**Comment by yangty89** [comment]
Hi @mp911de, I've done a detailed benchmarking comparing the performance of Jedis and Lettuce, in different mode. And I found that we got the highest score when using a list of Lettuce connections to serve requests under high concurrency. The scores is shown below. I think maybe we could introduce this mode in Lettuce :)

![image](https://user-images.githubusercontent.com/65708645/180597812-d9e6921d-c87f-4ac9-a903-ea6219ded113.png)
